lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -2011,15 +2014,21 @@ static void checkProtocolRefinementRequirements(ProtocolDecl *proto) {
     if (otherProto == proto)
       continue;
 
-    if (EnabledNoncopyableGenerics) {
+    if (EnabledNoncopyableGenerics
+        && otherProto->isSpecificProtocol(KnownProtocolKind::Copyable)) {
       // For any protocol 'P', there is an implied requirement 'Self: Copyable',
       // unless it was suppressed via `Self: ~Copyable`. So if this suppression
       // annotation exists yet Copyable was implied anyway, emit a diagnostic.
-      if (otherProto->isSpecificProtocol(KnownProtocolKind::Copyable))
-        if (!proto->canBeNoncopyable())
-          continue; // no ~Copyable annotation
-
-      // TODO(kavon): emit tailored error diagnostic to remove the ~Copyable
+      auto inverse = proto->getNoncopyableMarking().getInverse();
+      if (!inverse.isPresent())
+        continue; // no ~Copyable annotation
+
+      auto &Diags = proto->getASTContext().Diags;
+      Diags.diagnose(inverse.getLoc(),
+                     diag::noncopyable_generic_but_copyable,
+                     proto->getSelfInterfaceType(),
+                     getProtocolName(KnownProtocolKind::Copyable));
+      continue;
```

Comment:
Why isn't this diagnosed in diagnoseRequirementErrors() along with the other cases?

Reply:
I think it's because the StructuralRequirementsRequest isn't carrying the InverseRequirements up to RequirementSignatureRequest. I'll try your suggestion and see if that makes this bit of code redundant.

Reply:
Tried it and got a verification failure in the requirement machine. Will have to dig into it later.